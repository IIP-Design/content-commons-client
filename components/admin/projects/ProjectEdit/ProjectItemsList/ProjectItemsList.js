/**
 *
 * ProjectItemsList
 *
 */

import React from 'react';
import {
  bool, func, object, string
} from 'prop-types';
import gql from 'graphql-tag';
import { graphql } from 'react-apollo';
import orderBy from 'lodash/orderBy';
import { Icon, Loader } from 'semantic-ui-react';

import ProjectItem from 'components/admin/ProjectItem/ProjectItem';
import './ProjectItemsList.scss';

const ProjectItemsList = props => {
  const {
    listEl,
    data: { error, loading, project },
    projectId,
    headline,
    hasSubmittedData,
    projectType,
    displayItemInModal,
    modalTrigger,
    modalContent,
    customListStyle,
    customPlaceholderStyle
  } = props;

  const List = listEl;

  const defaultListStyle = {
    display: 'grid',
    gridTemplateColumns: 'repeat(auto-fill, minmax(13em, 1fr))',
    gridGap: '1.5em 1em',
    paddingLeft: '0',
    listStyle: 'none'
  };

  const listStyle = { ...defaultListStyle, ...customListStyle };

  if ( loading ) {
    return (
      <div style={ {
        display: 'flex',
        flexDirection: 'column',
        alignItems: 'center',
        justifyContent: 'center',
        minHeight: '200px'
      } }
      >
        <Loader
          active
          inline="centered"
          style={ { marginBottom: '1em' } }
          content="Loading the project items..."
        />
      </div>
    );
  }

  if ( error ) {
    return (
      <div className="project-items-list error">
        <p>
          <Icon color="red" name="exclamation triangle" />
          <span>Loading error</span>
        </p>
      </div>
    );
  }

  if ( !project ) return null;

  const { units } = project;

  const sortedUnits = orderBy( units, ['language.displayName', 'title'] );

  return (
    <div className="project-items">
      <h3 className="list-heading">{ headline }</h3>
      <List className="items-list" style={ listStyle }>
        { sortedUnits.map( unit => (
          <ProjectItem
            key={ `${unit.title} - ${unit.language.languageCode}` }
            isAvailable={ hasSubmittedData }
            type={ projectType }
            displayItemInModal={ displayItemInModal }
            projectId={ projectId }
            itemId={ unit.id }
            modalTrigger={ modalTrigger }
            modalContent={ modalContent }
            customPlaceholderStyle={ customPlaceholderStyle }
          />
        ) ) }
      </List>
    </div>
  );
};

ProjectItemsList.propTypes = {
  listEl: string,
  data: object.isRequired,
  projectId: string.isRequired,
  headline: string,
  hasSubmittedData: bool,
  projectType: string.isRequired,
  displayItemInModal: bool,
  modalTrigger: func,
  modalContent: func,
  customListStyle: object,
  customPlaceholderStyle: object
};

ProjectItemsList.defaultProps = {
  listEl: 'ul'
};

const PROJECT_ITEMS_QUERY = gql`
  query ProjectItems($id: ID!) {
    project: videoProject(id: $id) {
      units {
        id
        title
        language {
          displayName
        }
      }
    }
  }
`;

export default graphql( PROJECT_ITEMS_QUERY, {
  options: props => ( {
    variables: {
      id: props.projectId
    },
  } )
} )( ProjectItemsList );

export { PROJECT_ITEMS_QUERY };
