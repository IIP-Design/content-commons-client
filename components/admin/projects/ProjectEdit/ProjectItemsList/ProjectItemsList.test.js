import { mount, shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';
import wait from 'waait';
import { MockedProvider } from 'react-apollo/test-utils';
import { Icon, Loader } from 'semantic-ui-react';
import ProjectItemsList, { PROJECT_ITEMS_QUERY } from './ProjectItemsList';

const Trigger = () => <div className="trigger">Trigger</div>;
const Content = () => <div className="content">Content</div>;

const props = {
  listEl: 'ul',
  projectId: '123',
  headline: 'Videos in Project',
  hasSubmittedData: true,
  projectType: 'video',
  displayItemInModal: true,
  modalTrigger: Trigger,
  modalContent: Content
};

const mocks = [
  {
    request: {
      query: PROJECT_ITEMS_QUERY,
      variables: { id: props.projectId }
    },
    result: {
      data: {
        project: {
          units: [
            {
              id: 'edf123',
              title: 'Test Title YYY',
              language: {
                displayName: 'English'
              }
            },
            {
              id: 'mno456',
              title: 'Test Title ZZZ',
              language: {
                displayName: 'French'
              }
            }
          ]
        }
      }
    }
  }
];

const Component = (
  <MockedProvider mocks={ mocks } addTypename={ false }>
    <ProjectItemsList { ...props } />
  </MockedProvider>
);

describe( '<ProjectItemsList />', () => {
  it( 'renders initial loading state without crashing', () => {
    const wrapper = mount( Component );
    const projectItemsList = wrapper.find( 'ProjectItemsList' );
    const loader = (
      <Loader
        active
        inline="centered"
        style={ { marginBottom: '1em' } }
        content="Loading the project items..."
      />
    );

    expect( projectItemsList.exists() ).toEqual( true );
    expect( projectItemsList.contains( loader ) ).toEqual( true );
    expect( toJSON( projectItemsList ) ).toMatchSnapshot();
  } );

  it( 'renders error message if error is thrown', async () => {
    const errorMocks = [
      {
        request: {
          query: PROJECT_ITEMS_QUERY,
          variables: { id: props.projectId }
        },
        result: {
          errors: [{ message: 'There was an error.' }],
        }
      }
    ];

    const wrapper = mount(
      <MockedProvider mocks={ errorMocks } addTypename={ false }>
        <ProjectItemsList { ...props } />
      </MockedProvider>
    );
    // wait for the data and !loading
    await wait( 0 );
    wrapper.update();

    const projectItemsList = wrapper.find( 'ProjectItemsList' );
    const div = projectItemsList.find( 'div.project-items-list.error' );
    const icon = <Icon color="red" name="exclamation triangle" />;
    const span = <span>Loading error</span>;

    expect( projectItemsList.exists() ).toEqual( true );
    expect( div.exists() ).toEqual( true );
    expect( projectItemsList.contains( icon ) ).toEqual( true );
    expect( projectItemsList.contains( span ) ).toEqual( true );
  } );

  it( 'renders null if `project` is null', async () => {
    const nullMocks = [
      {
        request: {
          query: PROJECT_ITEMS_QUERY,
          variables: { id: props.projectId }
        },
        result: {
          data: { project: null },
        }
      }
    ];

    const wrapper = mount(
      <MockedProvider mocks={ nullMocks } addTypename={ false }>
        <ProjectItemsList { ...props } />
      </MockedProvider>
    );
    await wait( 0 );
    wrapper.update();

    const projectItemsList = wrapper.find( 'ProjectItemsList' );

    expect( projectItemsList.html() ).toEqual( null );
  } );

  it( 'renders the correct number of ProjectItem components', async () => {
    const wrapper = mount( Component );
    await wait( 0 );
    wrapper.update();

    const projectItemsList = wrapper.find( 'ProjectItemsList' );
    const projectItems = projectItemsList.find( 'ProjectItem' );
    const { units } = projectItemsList.prop( 'data' ).project;

    expect( projectItems ).toHaveLength( units.length );
  } );

  it( 'renders a custom element', async () => {
    const customEl = { listEl: 'ol' };

    const wrapper = mount(
      <MockedProvider mocks={ mocks } addTypename={ false }>
        <ProjectItemsList { ...{ ...props, ...customEl } } />
      </MockedProvider>
    );
    await wait( 0 );
    wrapper.update();

    const projectItemsList = wrapper.find( 'ProjectItemsList' );
    const ul = projectItemsList.find( 'ul' );
    const ol = projectItemsList.find( 'ol' );

    expect( ul.exists() ).toEqual( false );
    expect( ol.exists() ).toEqual( true );
  } );

  it( 'receives the Trigger component as a prop', () => {
    const wrapper = shallow( Component );
    const projectItems = wrapper.find( 'ProjectItem' );
    projectItems.forEach( item => {
      expect( item.prop( 'modalTrigger' )() )
        .toEqual( Trigger() );
    } );
  } );

  it( 'receives the Content component as a prop', () => {
    const wrapper = shallow( Component );
    const projectItems = wrapper.find( 'ProjectItem' );
    projectItems.forEach( item => {
      expect( item.prop( 'modalContent' )() )
        .toEqual( Content() );
    } );
  } );

  it( 'receives the projectType as a prop', () => {
    const wrapper = shallow( Component );
    const projectItems = wrapper.find( 'ProjectItem' );
    projectItems.forEach( item => {
      expect( item.prop( 'type' ) )
        .toEqual( props.projectType );
    } );
  } );

  it( 'receives the projectId as a prop', () => {
    const wrapper = shallow( Component );
    const projectItems = wrapper.find( 'ProjectItem' );
    projectItems.forEach( item => {
      expect( item.prop( 'projectId' ) )
        .toEqual( props.projectId );
    } );
  } );
} );
