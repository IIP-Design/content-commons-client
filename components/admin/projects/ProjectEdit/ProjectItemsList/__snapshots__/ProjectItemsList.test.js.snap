// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProjectItemsList /> renders initial loading state without crashing 1`] = `
<ProjectItemsList
  data={
    Object {
      "error": undefined,
      "fetchMore": [Function],
      "loading": true,
      "networkStatus": 1,
      "refetch": [Function],
      "startPolling": [Function],
      "stopPolling": [Function],
      "subscribeToMore": [Function],
      "updateQuery": [Function],
      "variables": Object {
        "id": "123",
      },
    }
  }
  displayItemInModal={true}
  hasSubmittedData={true}
  headline="Videos in Project"
  listEl="ul"
  modalContent={[Function]}
  modalTrigger={[Function]}
  projectId="123"
  projectType="video"
>
  <div
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
        "flexDirection": "column",
        "justifyContent": "center",
        "minHeight": "200px",
      }
    }
  >
    <Loader
      active={true}
      content="Loading the project items..."
      inline="centered"
      style={
        Object {
          "marginBottom": "1em",
        }
      }
    >
      <div
        className="ui active text centered inline loader"
        style={
          Object {
            "marginBottom": "1em",
          }
        }
      >
        Loading the project items...
      </div>
    </Loader>
  </div>
</ProjectItemsList>
`;
